userchroot.c

```diff
@@ -352,37 +351,91 @@ static int child_fn(void* v) {
   //
   // Mount the child's view of proc, which includes only processes
   // in its namespace.
-  rc = mkdir("/proc", S_IRWXU);
-  if(0 != rc && EEXIST != errno) {
+  rc = mount(
+             "proc",
+             "/proc",
+             "proc",
+             MS_REC|MS_NOSUID|MS_NODEV|MS_NOEXEC,
+             NULL
+             );
+  if(0 != rc) {
     fprintf(
             stderr,
-            "Failed to mkdir /proc. Error: %s\n", strerror(errno)
+            "Failed to mount proc. Error: %s\n", strerror(errno)
             );
     return rc;
   }
-  else {
-    rc = mount(
-               "proc",
-               "/proc",
-               "proc",
-               MS_REC|MS_NOSUID|MS_NODEV|MS_NOEXEC,
-               NULL
-               );
-    if(0 != rc) {
-      fprintf(
-              stderr,
-              "Failed to mount proc. Error: %s\n", strerror(errno)
-              );
-      return rc;
-    }
-  }
 #endif
 
   epilogue(ed);
   return 0;
 }
 #endif
 
+// Create a directory within the chroot path with the specified mode
+// intended to be used as a mount point.
+int create_mount_directory(const char* chroot_path,
+                          const char* directory,
+                          const mode_t perms) {
+  int dir_path_len = strlen(chroot_path) + strlen(directory) + 1;
+  char* dir_path = malloc(dir_path_len);
+  if (dir_path == NULL) {
+    fprintf(stderr,"Failed to allocate memory.\n");
+    return 1;
+  }
+  int rc = snprintf(dir_path, dir_path_len, "%s%s", chroot_path, directory);
+  if (rc <= 0) {
+    fprintf(stderr,"Failed to assemble path.\n");
+    free(dir_path);
+    return 1;
+  }
+  rc = mkdir(dir_path, perms);
+  if (rc < 0 && errno != EEXIST) {
+    fprintf(stderr, "Failed to mkdir %s. Error: %s\n", dir_path, strerror(errno));
+    free(dir_path);
+    return 1;
+  }
+  free(dir_path);
+  return 0;
+}
+
+int remove_mount_directory(const char* chroot_path,
+                          const char* directory) {
+  int dir_path_len = strlen(chroot_path) + strlen(directory) + 1;
+  char* dir_path = malloc(dir_path_len);
+  if (dir_path == NULL) {
+    fprintf(stderr,"Failed to allocate memory.\n");
+    return 1;
+  }
+  int rc = snprintf(dir_path, dir_path_len, "%s%s", chroot_path, directory);
+  if (rc <= 0) {
+    fprintf(stderr,"Failed to assemble path.\n");
+    free(dir_path);
+    return 1;
+  }
+  // Make sure this directory is st_uid to guard against removing
+  // directories userchroot itself didn't create
```

Comment:
Nit: The sentence is missing something as `st_uid` is an attribute/field, not an adjective. Maybe 'owned by root' or 'st_uid is 0'.